pyi codegen - removing byte-for-byte compatibility hacks

bdhirsh opened the pull request:
This PR alters the type signatures of some of the pyi type hints. Specifically, it modifies signatures in these files:
* /torch/_C/\_\_init\_\_.pyi
* torch/_C/_VariableFunctions.pyi
* torch/_VF.pyi

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I rationalized my way through the changes through a combination of
(a) staring at a few examples of function signature diffs and reasoning that the old version was incorrect
(b) letting CI pass

I can post a few more examples of differences (although posting the entire diff probably isn't worth it). Some examples, with their corresponding native_functions.yaml definitions:

```
-def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: randn.generator(int[] size, *, Generator? generator, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Optional[Tensor]=None) -> Tuple[Tensor, Tensor]: ...
+def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Union[Tensor, Tuple[Tensor, ...], List[Tensor]]=None) -> Tuple[Tensor, Tensor]: ...

- func: _mode.values(Tensor self, int dim=-1, bool keepdim=False, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
```

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49057 pyi codegen refactor - no need to group python signatures by overload name
* #49056 pyi codegen - removing byte-for-byte-compatibility hacks (sorting overloads)
* **#49055 pyi codegen - removing byte-for-byte compatibility hacks**
* #49054 pyi cleanup

Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

bdhirsh committed:
pyi codegen - removing byte-for-byte compatibility hacks

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen - removing byte-for-byte compatibility hacks"

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.


Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen - removing byte-for-byte compatibility hacks"

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.


Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

[ghstack-poisoned]

bdhirsh commented on tools/codegen/api/python.py:
```diff
@@ -753,12 +736,13 @@ def signature(f: NativeFunction, *, method: bool = False, pyi: bool = False) ->
 
 # TODO blowtorch
+# note: removing this will be BC-breaking. A quick test shows that
+# randperm will otherwise default its dtype to torch.float64
```
See the note. I tried removing, but it causes some tests to fail, which is why I noticed. We could either leave this in here or provide a deprecation warning and eventually take it out

ezyang replied:
This isn't actually pyi specific, right? This controls how we actually generate bindings

bdhirsh replied:
yeah that's right, Jiakai added this when he migrated python bindings. pyi doesn't actually touch it (another reason to not include in this PR, since I'm trying to only remove pya-specific hacks)

bdhirsh committed:
Update on "pyi codegen - removing byte-for-byte compatibility hacks"

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I rationalized my way through the changes through a combination of
(a) staring at a few examples of function signature diffs and reasoning that the old version was incorrect
(b) letting CI pass

I can post a few more examples of differences (although posting the entire diff probably isn't worth it). Some examples, with their corresponding native_functions.yaml definitions:

```
-def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: randn.generator(int[] size, *, Generator? generator, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Optional[Tensor]=None) -> Tuple[Tensor, Tensor]: ...
+def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Union[Tensor, Tuple[Tensor, ...], List[Tensor]]=None) -> Tuple[Tensor, Tensor]: ...

- func: _mode.values(Tensor self, int dim=-1, bool keepdim=False, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
```

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.


Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

[ghstack-poisoned]

ljk53 reviewed (approved):
LGTM
Procedure-wise, do we need tag PRs that change user facing interface so that it's highlighted in the release note?

bdhirsh commented:
> LGTM
> Procedure-wise, do we need tag PRs that change user facing interface so that it's highlighted in the release note?

I'm not aware of a specific tag (GitHub label?). Do you know where I can look to see if one exists?
I listed a little more clearly what the user-facing change is in the description.

ezyang commented:
This looks like a bugfix to me, no need to tag, it'll get rolled up when release notes get done.

bdhirsh committed:
Update on "pyi codegen - removing byte-for-byte compatibility hacks"

This PR alters the type signatures of some of the pyi type hints. Specifically, it modifies signatures in these files:
* /torch/_C/\_\_init\_\_.pyi
* torch/_C/_VariableFunctions.pyi
* torch/_VF.pyi

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I rationalized my way through the changes through a combination of
(a) staring at a few examples of function signature diffs and reasoning that the old version was incorrect
(b) letting CI pass

I can post a few more examples of differences (although posting the entire diff probably isn't worth it). Some examples, with their corresponding native_functions.yaml definitions:

```
-def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: randn.generator(int[] size, *, Generator? generator, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Optional[Tensor]=None) -> Tuple[Tensor, Tensor]: ...
+def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Union[Tensor, Tuple[Tensor, ...], List[Tensor]]=None) -> Tuple[Tensor, Tensor]: ...

- func: _mode.values(Tensor self, int dim=-1, bool keepdim=False, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
```

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.


Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

[ghstack-poisoned]

bdhirsh commented on tools/codegen/api/python.py:
```diff
@@ -710,11 +698,9 @@ def signature(f: NativeFunction, *, method: bool = False, pyi: bool = False) ->
             name='layout',
-            type=BaseType(BaseTy.Layout) if pyi else OptionalType(BaseType(BaseTy.Layout)),
+            type=OptionalType(BaseType(BaseTy.Layout)),
             default='strided' if pyi else 'torch.strided',
```
On second look, I think python binding code is actually wrong here and pyi is right (it should be `strided`).

All other uses of classes in the torch namespace are explicitly imported into the pyi file. For example, we use `Tensor` and not `torch.Tensor` throughout both the pyi and python binding code.

The python_arg_parser is just hardcoded to pick up `torch.strided`, so it would be a pretty easy change to unify them to use `strided`. Probably best for a separate PR that cleans up python binding hacks though.

bdhirsh committed:
Update on "pyi codegen - removing byte-for-byte compatibility hacks"

This PR alters the type signatures of some of the pyi type hints. Specifically, it modifies signatures in these files:
* /torch/_C/\_\_init\_\_.pyi
* torch/_C/_VariableFunctions.pyi
* torch/_VF.pyi

Removed the majority of the TODO hacks that I added to the original pyi PR to maintain byte-for-byte compatibility.

I rationalized my way through the changes through a combination of
(a) staring at a few examples of function signature diffs and reasoning that the old version was incorrect
(b) letting CI pass

I can post a few more examples of differences (although posting the entire diff probably isn't worth it). Some examples, with their corresponding native_functions.yaml definitions:

```
-def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def zeros(size: _size, *, names: Optional[Sequence[Union[str, ellipsis, None]]], dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: _layout=strided, device: Union[_device, str, None]=None, requires_grad:_bool=False) -> Tensor: ...
+def randn(*size: _int, generator: Optional[Generator], out: Optional[Tensor]=None, dtype: _dtype=None, layout: Optional[_layout]=torch.strided, device: Union[_device, str, None]=None, pin_memory: _bool=False, requires_grad: _bool=False) -> Tensor: ...

- func: randn.generator(int[] size, *, Generator? generator, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```

```
-def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Optional[Tensor]=None) -> Tuple[Tensor, Tensor]: ...
+def _mode(input: Tensor, dim: _int=-1, keepdim: _bool=False, *, out: Union[Tensor, Tuple[Tensor, ...], List[Tensor]]=None) -> Tuple[Tensor, Tensor]: ...

- func: _mode.values(Tensor self, int dim=-1, bool keepdim=False, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
```

I left a few of the divergences between pyi deprecated vs. native signatures, since (a) they're smaller and (b) it might make more sense to kill the deprecated functions at some point entirely.


Differential Revision: [D25410847](https://our.internmc.facebook.com/intern/diff/D25410847)

[ghstack-poisoned]
